leshan-core/src/main/java/org/eclipse/leshan/core/attributes/AttributeSet.java

```diff
@@ -0,0 +1,144 @@
+        Attribute pmax = attributeMap.get(AttributeName.MAXIMUM_PERIOD);
+        if ((pmin != null) && (pmax != null) && (Long) pmin.getValue() > (Long) pmax.getValue()) {
+            throw new IllegalArgumentException(String.format("Cannot write attributes where '%s' > '%s'",
+                    pmin.getCoRELinkParam(), pmax.getCoRELinkParam()));
+        }
+    }
+
+
+    /**
+     * Returns a new AttributeSet, containing only the attributes that have a matching
+     * Attachment level. 
+     * @param attachment the Attachment level to filter by
+     * @return a new {@link AttributeSet} containing the filtered attributes
+     */
+    public AttributeSet filter(Attachment attachment) {
+        List<Attribute> attrs = new ArrayList<>();
+        for (Attribute attr : getAttributes()) {
+            if (attr.getAttachment() == attachment) {
+                attrs.add(attr);
+            }
+        }
+        return new AttributeSet(attrs.toArray(new Attribute[attrs.size()]));
+    }
+    
+    /**
+     * Creates a new AttributeSet by merging another AttributeSet onto this instance.
+     * @param attributes the AttributeSet that should be merged onto this instance. Attributes in this
+     * set will overwrite existing attribute values, if present. If this is null, the
+     * new attribute set will effectively be a clone of the existing one
+     * @return the merged AttributeSet 
+     */
+    public AttributeSet merge(AttributeSet attributes) {
+        Map<AttributeName, Attribute> merged = new LinkedHashMap<>();
+        for (Attribute attr : getAttributes()) {
+            merged.put(attr.getName(), attr);
+        }
+        if (attributes != null) {
+            for (Attribute attr : attributes.getAttributes()) {
+                merged.put(attr.getName(), attr);
+            }
+        }
+        return new AttributeSet(merged.values().toArray(new Attribute[merged.size()]));        
+    }
+
+    /**
+     * Returns the attributes as a map with the CoRELinkParam as key and the attribute value as map value.
+     * @return the attributes map
+     */
+    public Map<String, Object> getMap() {
+        Map<String, Object> result = new LinkedHashMap<>();
+        for (Attribute attr : attributeMap.values()) {
+            result.put(attr.getCoRELinkParam(), attr.getValue());
+        }
+        return result;
+    }
+    
+    public Attribute[] getAttributes() {
+        return attributeMap.values().toArray(new Attribute[attributeMap.size()]);
+    }
+    
+    public String[] toQueryParams() {
+        List<String> queries = new LinkedList<>();
+        for (Attribute attr : attributeMap.values()) {
+            queries.add(String.format("%s=%s", attr.getCoRELinkParam(), attr.getValue()));
+        }
+        return queries.toArray(new String[queries.size()]);
+    }
+
+    @Override
+    public String toString() {
+        return String.join("&", toQueryParams()); 
```

Comment:
We aims java 1.7 (mainly because of android)
String.join is available since java 1.8
I should change the maven configuration to check that.